File: packages/android-views/src/main/kotlin/org/digitalgreen/farmerchat/views/network/GuestApiClient.kt
package org.digitalgreen.farmerchat.views.network

import android.util.Log
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONObject
import java.io.IOException
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

/**
 * Standalone client used exclusively for guest user initialisation.
 *
 * Uses a plain `API-Key` header (not Bearer JWT).
 * Endpoint: `POST /api/user/initialize_user/`
 */
internal class GuestApiClient(private val baseUrl: String) {

    private companion object {
        const val TAG = "FC.GuestApiClient"
        const val GUEST_API_KEY = "Y2K3kW5R9uQ0fL2X8zI7hT3aJ7"
        const val PATH_INITIALIZE = "api/user/initialize_user/"
        const val TIMEOUT_MS = 30_000
    }

    suspend fun initializeUser(deviceId: String): InitializeGuestUserResponse =
        withContext(Dispatchers.IO) {
            var connection: HttpURLConnection? = null
            try {
                val url = URL("${baseUrl.trimEnd('/')}/$PATH_INITIALIZE")
                Log.d(TAG, "→ POST $url")
                connection = (url.openConnection() as HttpURLConnection).apply {
                    requestMethod = "POST"
                    connectTimeout = TIMEOUT_MS
                    readTimeout = TIMEOUT_MS
                    doOutput = true
                    setRequestProperty("Content-Type", "application/json")
                    setRequestProperty("API-Key", GUEST_API_KEY)
                }

                val body = JSONObject().apply { put("device_id", deviceId) }.toString()
                OutputStreamWriter(connection.outputStream, Charsets.UTF_8).use { w ->
                    w.write(body); w.flush()
                }

                val code = connection.responseCode
                if (code !in 200..299) {
                    val error = connection.errorStream
                        ?.bufferedReader(Charsets.UTF_8)?.use { it.readText() }.orEmpty()
                    throw ApiException(code, error)
                }

                val text = connection.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
                val json = JSONObject(text)

                val response = InitializeGuestUserResponse(
                    accessToken = json.getString("access_token"),
                    refreshToken = json.getString("refresh_token"),
                    userId = json.optString("user_id", ""),
                    createdNow = json.optBoolean("created_now", false),
                    showCropsLivestocks = json.optBoolean("show_crops_livestocks", false),
                    countryCode = json.optString("country_code", null),
                    country = json.optString("country", null),
                    state = json.optString("state", null),
                )

                TokenStore.saveTokens(
                    accessToken  = response.accessToken,
                    refreshToken = response.refreshToken,
                    userId       = response.userId,
                    countryCode  = response.countryCode,
                    country      = response.country,
                    state        = response.state,
                )

                Log.d(TAG, "Guest user initialized. userId=${response.userId} countryCode=${response.countryCode}")
                response
            } finally {
                connection?.disconnect()
            }
        }
}

File: packages/ios-swiftui/Sources/FarmerChatSwiftUI/Network/GuestAPIClient.swift
import Foundation
import os

/// Standalone client used exclusively for guest user initialisation.
///
/// Uses `API-Key: <key>` header — the only endpoint
/// that does not require a prior JWT token.
///
/// Endpoint: `POST /api/user/initialize_user/`
struct GuestAPIClient {

    private static let logger = Logger(
        subsystem: "org.digitalgreen.farmerchat",
        category: "GuestAPIClient"
    )

    static let guestApiKey = "Y2K3kW5R9uQ0fL2X8zI7hT3aJ7"

    private let baseUrl: String

    init(baseUrl: String) {
        self.baseUrl = baseUrl.trimmingCharacters(in: .init(charactersIn: "/"))
    }

    /// Initialize (or re-identify) a guest user by device ID.
    ///
    /// On success, tokens are saved to `TokenStore.shared`.
    ///
    /// - Parameter deviceId: Stable device UUID from `DeviceInfoProvider.stableDeviceId()`.
    @discardableResult
    func initializeUser(deviceId: String) async throws -> InitializeGuestUserResponse {
        guard let url = URL(string: "\(baseUrl)/api/user/initialize_user/") else {
            throw NetworkError.invalidURL
        }

        var request = URLRequest(url: url, timeoutInterval: 30)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(Self.guestApiKey, forHTTPHeaderField: "API-Key")
        request.httpBody = try JSONEncoder().encode(["device_id": deviceId])

        let (data, response) = try await URLSession.shared.data(for: request)

        guard let http = response as? HTTPURLResponse else {
            throw NetworkError.unknown("Invalid response")
        }
        guard (200...299).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? ""
            throw NetworkError.serverError(http.statusCode, body)
        }

        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        let parsed = try decoder.decode(InitializeGuestUserResponse.self, from: data)

        await TokenStore.shared.saveTokens(
            accessToken: parsed.accessToken,
            refreshToken: parsed.refreshToken,
            userId: parsed.userId ?? "",
            countryCode: parsed.countryCode,
            country: parsed.country,
            state: parsed.state
        )

        Self.logger.debug("Guest user initialized. userId=\(parsed.userId ?? "nil")")
        return parsed
    }
}
